cpp/include/cudf/io/detail/orc.hpp

```diff
@@ -62,15 +67,68 @@ class reader {
   /**
    * @brief Destructor explicitly declared to avoid inlining in header
    */
-  ~reader();
+  virtual ~reader();
 
   /**
    * @brief Reads the entire dataset.
    *
-   * @param options Settings for controlling reading behavior
    * @return The set of columns along with table metadata
    */
-  table_with_metadata read(orc_reader_options const& options);
+  table_with_metadata read();
+};
+
+/**
+ * @brief The reader class that supports iterative reading from an array of data sources.
+ *
+ * This class intentionally subclasses the `reader` class with private inheritance to hide the
+ * base class `reader::read()` API. As such, only chunked reading APIs are supported through it.
+ */
+class chunked_reader : private reader {
```

Comment:
why do we derive from `reader`? looks like we only use `_impl`. We could just have `std::unique_ptr<impl> _impl;` data member in this class instead.

Reply:
Sounds good, I'll modify that.

Reply:
Done in https://github.com/rapidsai/cudf/pull/15094/commits/734dcf363ee2c7a5fcf91e311913f6b44933440d.